### ISSUE TITLE
Phpactor constantly busy altough no significant changes were made

### ISSUE THREAD

**Issue by BladeMF** [phpactor/phpactor#2394]
Sorry about the title, but I am not able to describe this in one sentence more eloquently. 
`Phpactor` seems to be constantly busy resolving stuff (the log keeps showing "Resolving...") altough currently the only thing I did was create an empty file and requested the code actions on it. This takes **minutes** every time. There seems to be a moment when it stops for a while, but it is not clear to me that causes it, why does it keep parsing and resolving stuff. Only once I saw a message saying it killed some process because it took more than 5 seconds. 

I am not sure now to debug this. All the while I was writing this it kept parsing an resolving although I am not doing anything. It seems just as long as if it is parsing the whole codebase.

Any suggestions?

**Comment by BladeMF** [comment]
Oh, and as a result I get a bunch of hung "Resolving code actions..." messages in the IDE.

**Comment by ging-dev** [comment]
> Oh, and as a result I get a bunch of hung "Resolving code actions..." messages in the IDE.

Maybe it duplicate of https://github.com/phpactor/phpactor/issues/2381#issuecomment-1732549899

**Comment by nrob81** [comment]
I had a similar problem, I was flooded with "[Resolving code actions]" status messages and my IDE (nvim) was totally slowed  down.
@BladeMF i disabled lightbulb and now it is ok. 

my config:
```lua
config = function()
    require("lspsaga").setup({
        outline = {
            auto_preview = false
        },
        lightbulb = {
            enable = true,
        }
    })
end,
```

**Comment by BladeMF** [comment]
I'd rather understand what is causing it. It seems connected to browsing files in the `vendors` folder.

**Comment by dantleech** [comment]
creating a reproduction repo would help. I can't reproduce this behavior....

**Comment by BladeMF** [comment]
I am not sure this is vscode specific. The log says endless resolutions of the same classes and functions, over and over again. If you wait long enough they all finish eventually, but we are talking about multiple minutes here.

@dantleech do you have ideas how I can debug this further?

**Comment by dantleech** [comment]
create a reproduction repository? enable LSP tracing configuration option to see what (if anything) VS code is requeseting from Phpactor?

**Comment by BladeMF** [comment]
I wonder if these resolutions could be cached within the request, i.e. when a symbol is resolved in the scope of a single request, it should not be resolved again. Or should we record which provider resolved it so then just that provider is called or something.

**Comment by dantleech** [comment]
they are cached IIRC, the logs show the cached resolutions as well as the non-cached ones.

**Comment by BladeMF** [comment]
1. `IIRC` = ?
2. How should a cached resolution look like? Nothing stands out when I read the logs
3. Yes, probably. I have currently set `language_server.diagnostics_on_update` to `false` and am waiting to see what happens. It may be that `phpactor` is just getting swamped.

**Comment by BladeMF** [comment]
```php
$this->logger->debug(sprintf('Resolving: %s', get_class($node)));
```
This means that if I see "Resolving" in the logs then the cache is not being used, correct? If true then a single "save" produces 390 lines of `[2024-01-11 09:30:54] phpactor.DEBUG: Resolving: Microsoft\PhpParser\Node\Expression\MemberAccessExpression {"channel":"wr"} []`. I assume there might be more than one requests sent per save, but still, they are not 390. Am I missing something?

**Comment by BladeMF** [comment]
Soo, a signle save on a file:

## Requests
```
[Trace - 10:30:55 AM] Sending notification 'textDocument/didSave'.
Params: {
    "textDocument": {
        "uri": "file:///var/www/ct/app/tests/UseCase/Integration/VerifyIssueAnalysisTest.php",
        "version": 28
    }
}


[Trace - 10:30:55 AM] Received notification 'textDocument/publishDiagnostics'.
Params: {
    "uri": "file:///var/www/ct/app/tests/UseCase/Integration/VerifyIssueAnalysisTest.php",
    "version": 1,
    "diagnostics": []
}


[Trace - 10:30:55 AM] Sending request 'textDocument/codeAction - (45)'.
Params: {
    "textDocument": {
        "uri": "file:///var/www/ct/app/tests/UseCase/Integration/VerifyIssueAnalysisTest.php"
    },
    "range": {
        "start": {
            "line": 437,
            "character": 55
        },
        "end": {
            "line": 437,
            "character": 55
        }
    },
    "context": {
        "diagnostics": []
    }
}


[Trace - 10:30:55 AM] Sending notification 'workspace/didChangeWatchedFiles'.
Params: {
    "changes": [
        {
            "uri": "file:///var/www/ct/app/tests/UseCase/Integration/VerifyIssueAnalysisTest.php",
            "type": 2
        }
    ]
}


[Trace - 10:30:56 AM] Received request 'window/workDoneProgress/create - (a7cb2305-046f-4a6f-a210-7e500e8d2e8d)'.
Params: {
    "token": "462edb2e-ceec-424b-81da-5cb0ac7c0262"
}


[Trace - 10:30:56 AM] Sending response 'window/workDoneProgress/create - (a7cb2305-046f-4a6f-a210-7e500e8d2e8d)'. Processing request took 0ms
No result returned.


[Trace - 10:30:56 AM] Received notification '$/progress'.
Params: {
    "token": "462edb2e-ceec-424b-81da-5cb0ac7c0262",
    "value": {
        "kind": "begin",
        "title": "Resolving code actions"
    }
}


[Trace - 10:31:00 AM] Sending notification '$/cancelRequest'.
Params: {
    "id": 45
}


[Trace - 10:31:03 AM] Received notification 'textDocument/publishDiagnostics'.
Params: {
    "uri": "file:///var/www/ct/app/tests/UseCase/Integration/VerifyIssueAnalysisTest.php",
    "version": 1,
    "diagnostics": [
        {
            "range": {
                "start": {
                    "line": 380,
                    "character": 1
                },
                "end": {
                    "line": 380,
                    "character": 100
                }
            },
            "severity": 1,
            "source": "phpstan",
            "message": "Call to deprecated method isComplete() of class MarsIPM\\Entity\\IssueAnalysis:\nuse status"
        },
        {
            "range": {
                "start": {
                    "line": 381,
                    "character": 1
                },
                "end": {
                    "line": 381,
                    "character": 100
                }
            },
            "severity": 1,
            "source": "phpstan",
            "message": "Call to deprecated method getCompletedByUser() of class MarsIPM\\Entity\\IssueAnalysis:\nuse validatedByUser"
        },
        {
            "range": {
                "start": {
                    "line": 382,
                    "character": 1
                },
                "end": {
                    "line": 382,
                    "character": 100
                }
            },
            "severity": 1,
            "source": "phpstan",
            "message": "Call to deprecated method getCompletedOnDate() of class MarsIPM\\Entity\\IssueAnalysis:\nuse validatedOnDate"
        },
        {
            "range": {
                "start": {
                    "line": 438,
                    "character": 1
                },
                "end": {
                    "line": 438,
                    "character": 100
                }
            },
            "severity": 1,
            "source": "phpstan",
            "message": "Call to deprecated method isComplete() of class MarsIPM\\Entity\\IssueAnalysis:\nuse status"
        },
        {
            "range": {
                "start": {
                    "line": 439,
                    "character": 1
                },
                "end": {
                    "line": 439,
                    "character": 100
                }
            },
            "severity": 1,
            "source": "phpstan",
            "message": "Call to deprecated method getCompletedByUser() of class MarsIPM\\Entity\\IssueAnalysis:\nuse validatedByUser"
        },
        {
            "range": {
                "start": {
                    "line": 440,
                    "character": 1
                },
                "end": {
                    "line": 440,
                    "character": 100
                }
            },
            "severity": 1,
            "source": "phpstan",
            "message": "Call to deprecated method getCompletedOnDate() of class MarsIPM\\Entity\\IssueAnalysis:\nuse validatedOnDate"
        }
    ]
}


[Trace - 10:31:19 AM] Received notification '$/progress'.
Params: {
    "token": "462edb2e-ceec-424b-81da-5cb0ac7c0262",
    "value": {
        "kind": "end"
    }
}


[Trace - 10:31:19 AM] Received response 'textDocument/codeAction - (45)' in 23353ms. Request failed: Request 45 (textDocument/codeAction) cancelled (-32800).
```
## Cache stats:
### Current cache implementation:
* 37,680 cache logs (I created them)
* 6,464 cache hits
* 30,362 cache misses
* 854 `StaticCache` objects created
### Permanent cache (just so we know what _could be_)
* 11,567 cache logs (I created them)
* 2,439 cache hits
* 8,823 cache misses

Not sure why a second save would not have only hits (since the cache is static property).

**Comment by dantleech** [comment]
> This means that if I see "Resolving" in the logs then the cache is not being used, correct

seems so.

the current cache implementation is necessary due to the way we analyse the code, so while "permanant" cache is faster, it will not work very well.

**Comment by BladeMF** [comment]
I figured as much. 

The interesting bit is that with normal cache there are _consistently_ 3 times as much calls (I also see two notifications for "Resolving code actions"). I double checked it. The test is really simple:
1. Wait for all activity to stop
2. Without the mouse being over the document, press Ctrl+S
3. Wait for all activity to stop
4. Copy the logs and count miss & hit
So I rule out that I am doing something that provokes more work from `phpactor`.

**Comment by BladeMF** [comment]
Otherwise it would appear that `language_server.diagnostics_on_update` is working out nicely for now.

**Comment by dantleech** [comment]
yes, as mentioned the analysis process is much slower than it could be, it's very complex and I haven't had the time or motivation to dive back into it :upside_down_face:

**Comment by BladeMF** [comment]
Yeah, I know how that feels.

**Comment by BladeMF** [comment]
So, I think I've figured it out. If I click around the file the requests just mount one on top of each other. I am trying to increase `language_server.diagnostic_sleep_time` now.

**Comment by dantleech** [comment]
the "outsourced" diagnostic provider should limit the number of concurrent processes, but maybe the bottleneck is in the main process.

**Comment by BladeMF** [comment]
Should it drop the surplus or should they just wait?

**Comment by dantleech** [comment]
AFAIK it should ensure there is only one process running at a time. can't find the code now.

**Comment by BladeMF** [comment]
So, some more data from today. I seem to be able to swamp it again (or at least _I think_ I am swamping it). Tens of requests for "resolving actions" are in queue for minutes (!) at a time. I've turned on profiling, but currently it is not obvious what is causing the delay.

**Comment by BladeMF** [comment]
Hmm
```
[2024-01-12 14:15:27] phpactor.INFO: PROF 695.7335 << code-action [ImportNameProvider] import unresolvable class names {"channel":"LSP"} []
[2024-01-12 14:15:27] phpactor.INFO: PROF 694.9455 << code-action [ImportNameProvider] import unresolvable class names {"channel":"LSP"} []
```

**Comment by BladeMF** [comment]
More examples:
```
[2024-01-12 14:20:32] phpactor.INFO: PROF 119.9579 << request #466 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 119.8677 << request #467 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 119.8661 << request #468 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 119.8652 << request #469 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 119.8638 << request #470 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 119.8626 << request #471 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 119.7820 << request #472 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 62.8641 << request #473 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 60.5340 << request #474 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 60.5333 << request #475 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 60.5326 << request #476 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 60.5247 << request #477 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 60.5242 << request #478 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 60.4189 << request #479 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 60.4172 << request #480 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 60.3820 << request #481 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 60.3811 << request #482 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 60.3668 << request #483 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 60.3662 << request #484 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 60.3642 << request #485 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 60.3545 << request #486 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 60.3540 << request #487 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 60.3530 << request #488 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 59.4077 << request #489 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 59.3554 << request #490 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 55.9894 << request #491 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 55.9891 << request #492 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 55.9885 << request #493 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 55.9749 << request #494 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 55.9384 << request #495 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 55.9378 << request #496 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 55.9372 << request #497 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 55.9364 << request #498 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 55.9356 << request #499 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 55.8879 << request #500 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 55.8873 << request #501 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 55.8863 << request #502 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 55.8774 << request #503 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 55.8769 << request #504 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 54.6000 << request #505 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 54.5989 << request #506 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 54.5980 << request #507 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 54.5560 << request #508 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 54.5550 << request #509 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 54.4577 << request #510 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 38.5521 << request #511 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 38.4928 << request #512 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 38.4918 << request #513 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 38.4913 << request #514 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 38.4906 << request #515 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 38.4904 << request #516 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 38.4899 << request #517 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 38.4894 << request #518 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:32] phpactor.INFO: PROF 38.4889 << request #519 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:34] phpactor.INFO: PROF 40.2546 << request #520 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:34] phpactor.INFO: PROF 40.2160 << request #521 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:34] phpactor.INFO: PROF 38.2071 << request #522 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:20:34] phpactor.INFO: PROF 35.9134 << request #523 [textDocument/codeAction] {"channel":"LSP"} []
```
I don't think that 5 second thing is happening.

**Comment by BladeMF** [comment]
How about those:
```
[2024-01-12 14:43:04] phpactor.INFO: PROF 1,830.1591 << code-action [ImportNameProvider] import unresolvable class names {"channel":"LSP"} []
[2024-01-12 14:43:04] phpactor.INFO: PROF 1,830.2361 << request #457 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:43:09] phpactor.INFO: PROF 1,798.2580 << code-action [ImportNameProvider] import unresolvable class names {"channel":"LSP"} []
[2024-01-12 14:43:09] phpactor.INFO: PROF 1,663.1889 << code-action [ImportNameProvider] import unresolvable class names {"channel":"LSP"} []
[2024-01-12 14:43:09] phpactor.INFO: PROF 1,571.8249 << code-action [ImportNameProvider] import unresolvable class names {"channel":"LSP"} []
[2024-01-12 14:43:09] phpactor.INFO: PROF 1,798.8652 << request #459 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:43:09] phpactor.INFO: PROF 1,573.2020 << request #461 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:43:09] phpactor.INFO: PROF 1,665.0321 << request #460 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:49:14] phpactor.INFO: PROF        >> request #526 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:49:14] phpactor.INFO: PROF        >> code-action [ReplaceQualifierWithImportProvider] replace qualifier with importer {"channel":"LSP"} []
[2024-01-12 14:49:14] phpactor.INFO: PROF 0.0209 << code-action [ReplaceQualifierWithImportProvider] replace qualifier with importer {"channel":"LSP"} []
[2024-01-12 14:49:14] phpactor.INFO: PROF        >> code-action [ImportNameProvider] import unresolvable class names {"channel":"LSP"} []
[2024-01-12 14:49:17] phpactor.INFO: PROF 1,723.3390 << code-action [ImportNameProvider] import unresolvable class names {"channel":"LSP"} []
[2024-01-12 14:49:17] phpactor.INFO: PROF 1,640.9769 << code-action [ImportNameProvider] import unresolvable class names {"channel":"LSP"} []
[2024-01-12 14:49:17] phpactor.INFO: PROF 3.4846 << code-action [ImportNameProvider] import unresolvable class names {"channel":"LSP"} []
[2024-01-12 14:49:17] phpactor.INFO: PROF        >> code-action [TransformerCodeActionPovider] "promote_constructor" transformer {"channel":"LSP"} []
[2024-01-12 14:49:19] phpactor.INFO: PROF 1.7076 << code-action [TransformerCodeActionPovider] "promote_constructor" transformer {"channel":"LSP"} []
[2024-01-12 14:49:19] phpactor.INFO: PROF 1,756.6959 << request #524 [textDocument/codeAction] {"channel":"LSP"} []
[2024-01-12 14:49:19] phpactor.INFO: PROF 1,642.7082 << request #525 [textDocument/codeAction] {"channel":"LSP"} []
```

I have no idea why `ImportNameProvider` chokes so bad. If you give ideas I can try and profile that thing in particular.

**Comment by BladeMF** [comment]
I've disabled everything that tries to find unresolvable classes.

**Comment by BladeMF** [comment]
And now everything that starts with `add_missing`. I _cannot_ understand why _this_ file causes this.

**Comment by BladeMF** [comment]
So, currently the problem seems to be gone. I disabled:
* ImportNameProvider
* add_missing_class_generic
* CreateUnresolvableClassProvider
* CorrectUndefinedVariableCodeAction
* add_missing_properties
* add_missing_docblocks_return
* add_missing_params
* add_missing_return_types
* remove_unused_imports
* GenerateMemberProvider

@dantleech I am guessing the common denominator is that these all enumerate all known classess or something. I'd love to profile this stuff.

**Comment by dantleech** [comment]
well they probably all analyse the entire AST. does the performance problems scale witrh the number of actions you enable? I'd imagine that one is basically as bad as all of them?

**Comment by BladeMF** [comment]
I think the first one does the work and the others have basically no effect. It's not like it scales. You know it's doing work if you see endless resolvings in the log. I have a faint memory trying to find the problem a while back and got nowhere or even worse, to some depressing conclusion (i.e. not easly solved). I might have a go again.

**Comment by BladeMF** [comment]
The problem is definitely gone. So reading the AST into memory seems to be the problem. I wonder if the pure number of classes is the problem or there is something else available that can help this (short of stopping the functionalities that use this).

**Comment by dantleech** [comment]
The AST is always cached, but it' invalidated. I assume you are talking about changing the cache key?

**Comment by BladeMF** [comment]
I am not really sure what am I talking about. What I imagine happens is that thit analysis takes so long that each subsequent pending analysis makes the situation somehow worse (that's why we are seeing these 1,6k seconds) as opposed to a consistent, say, 10 seconds or something. Probably the initial request is not cancelled and we end up running with 13 simultaneous request, degrading performance.

**Comment by axelrindle** [comment]
I recorded the following showing metrics of the LSP process:

https://github.com/phpactor/phpactor/assets/7105632/abbbe67e-67b3-4c92-8cc6-4625229a56b8

Whenever VSCode says **Resolving code actions** CPU usage goes up to 100% (on a single core) and the notification area starts stacking with these messages.

**Comment by claytonrcarter** [comment]
This is happening in Zed, too: https://github.com/zed-industries/zed/issues/20249

FWIW I am using `phpactor` in Laravel projects and it works great until this pops up. Most of time these resolve w/i a few seconds, but it's not uncommon for multiple requests to back up and cause the LSP magic to stop working. If I wait long enough, these will often resolve, but that can take several minutes, during which time the LSP is unresponsive in the editor. For example, as I write this (in Firefox, not in Zed :) I have watched a stack of 5-6 "Resolving code actions" requests crunch through and one by one resolve. I would estimate that this took ~2-3 minutes to clear out. It's not uncommon to see this stack get upwards of 10-12 pending requests.

Though it would be great to improve performance, would exposing a `language_server.code_actions_timeout` config work as a quick fix for this? There are already several of these timeouts available, but it looks like code actions aren't covered yet.

**Comment by MadeInShineA** [comment]
Hey I'm also doing some laravel work on zed and got the same issue, any update on this ?

**Comment by dantleech** [comment]
Hm, I don't notice this in Neovim as code actions are _requested_. Looking at VS code in a very complex source file though I can see the editor is requesting code actions after most changes:

<img width="661" height="621" alt="Image" src="https://github.com/user-attachments/assets/b6db0c88-a37a-4c57-b520-23a939101497" />

Code actions do currently take some time to resolve but they should not stack. I'll look into it.
